### PR TITLE
[WIP] Starting kit with image procesing approach

### PR DESCRIPTION
**Pull request by glemaitre**

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
If we see this as an example submission, I think the `extraction.py` should live in a directory in `submissions/..` ?

**Comment by glemaitre** [comment]
yep I moved it there now that I am plugging stuff.
However, I am still confused how to import iou from the local workflow since it is not installable for the moment. I don't think this is a concern for the moment. I just want to see the output and then fix those location and import issues.

**Comment by jorisvandenbossche** [comment]
No, for now you cannot import. I would say to just copy paste it for now, and we can try to think of other ways.

File: submissions/image_processing/object_detector.py [hunk 1]
```diff
@@ -1,59 +1,66 @@
...
-        self.threshold = threshold
+from .extraction import BlobExtractor
 
-    def fit(self, X, y):
-        return self
 
-    def _hough_detection(self, image):
-        edges = canny(image, sigma=self.sigma, low_threshold=10,
-                      high_threshold=50)
+class CraterDetector(object):
```
Comment (jorisvandenbossche) [inline submissions/image_processing/object_detector.py]: this has to be ObjectDetector

**Comment by glemaitre** [comment]
I obtain the following results with Zernike and SURF descriptor.

```python
import numpy as np

from sklearn.ensemble import GradientBoostingClassifier

from problem import get_train_data, get_test_data

from submissions.image_processing.object_detector import ObjectDetector
from submissions.image_processing.extraction import BlobExtractor

from workflow.scores import Precision, Recall

X_train, y_train = get_train_data()
X_train = np.array(X_train)
X_test, y_test = get_test_data()
X_test = np.array(X_test)

clf = ObjectDetector(extractor=BlobExtractor(iou_threshold=0.4, max_radius=30, overlap=0.2, padding=2.0),
                     estimator=GradientBoostingClassifier(n_estimators=100),
                     n_jobs=-1)

clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)

precision, recall = Precision(), Recall()
print('Precision: {} - Recall: {} '.format(precision(y_test, y_pred),
                                           recall(y_test, y_pred)))
```

```
Precision: 1.0 - Recall: 0.8275862068965517
```

@aboucaud @jorisvandenbossche Is it enough for the starting kit?

**Comment by glemaitre** [comment]
the failing test is something strange. I don't think flake8 is correct for this one.

**Comment by glemaitre** [comment]
Also the blob detector is not precise since it will detect white or black blob. the issue is that craters are white and black so my crater will always be shifted and thus the IOU will decrease.

So there is room to improve the border location in fact.

**Comment by kegl** [comment]
Can we merge this? I'd like to work on the score and I need a decent detector.

**Comment by aboucaud** [comment]
I thought GH had a possibility for a reviewer to push on the branch of the PR author..but apparently not. Or @glemaitre  has not allowed for that.

**Review by aboucaud — DISMISSED** [review]
Please add an `--ignore F812` after the `flake8` command in the `.travis.yml`

**Comment by kegl** [comment]
I just copy pasted now. I needed to add the confidences to the predictions, then it worked. On my data it doesn't perform very well:

----------------------------
test scp = 0.998 ± 0.0
test OSPA = 0.0064 ± 0.0
test average_precision = 0.0 ± 0.0
test precision = 0.0023 ± 0.0
test recall = 0.2644 ± 0.0

**Comment by aboucaud** [comment]
Passing now, shall I merge ?

**Comment by kegl** [comment]
LGTM

**Comment by glemaitre** [comment]
Arriving a bit late on this one. The only thing that can be done is to remove the notebook `segmentation.ipynb`. It will be useless since it was just a scratch book.

**Comment by glemaitre** [comment]
@kegl on which data did you try:

> I just copy pasted now. I needed to add the confidences to the predictions, then it worked. On my data it doesn't perform very well:

I was getting a precision of 0.7 and a recall of 0.3

**Comment by jorisvandenbossche** [comment]
Is this actually working? @glemaitre can you run `ramp_test_submissions --submission image_processing` with this? 
(I would have expected that there are import problems because of the multiple files)

**Comment by kegl** [comment]
Only @jorisvandenbossche knows :). It has ~4000 craters. The test is running on 200 patches.

**Comment by glemaitre** [comment]
yep but I did apply the patch that Joris gave me to test on the all data. However, not CV I assume.

**Comment by jorisvandenbossche** [comment]
So there are still some problems with this as a starting kit example (it is *not* running with `ramp_test_submission`).

The main reason is the 'import'-problem. So we have two problems:

- the fact that you cannot organize your code in multiple files (inside a submission dir), which this PR did. So we need to copy paste everything inside a single file. I personally don't really like this, but we won't change this for this ramp, so need to change this? (@glemaitre can you do this?) 
- the fact that you cannot import (and thus cannot reuse) things from the local workflow. Therefore this submission copy pasted code for calculating IOU, which is IMO not ideal. It would be nice to solve this for the ramp.

**Comment by glemaitre** [comment]
I am on it

**Comment by kegl** [comment]
The solution is to put the score and the workflow into ramp-workflow. That's the reason it exists.

On the multiple files issue, this will not change. In ramp-board we need to know the file names the participants will submit. There is also a deeper design issue here which you will understand if you read this book.

https://www.amazon.com/gp/product/0691148902/ref=as_li_tf_tl?ie=UTF8&tag=michaniels-20&linkCode=as2&camp=217145&creative=399373&creativeASIN=0691148902

Two of the three keys to successful crowdsourcing is modularization and allowing minimal contributions. The reason this is slightly messy here is because we don't yet know the detection workflow very well. Once we do, we can design a more granular standardized workflow (or even alternative workflows) with several files. Having a second detection task would also help, we'd see more easily what's common and what's moving.

**Comment by jorisvandenbossche** [comment]
> The solution is to put the score and the workflow into ramp-workflow. That's the reason it exists.

IMO this is not scalable on a long term (e.g. if the ambition is that multiple universities use it for courses with a variety of challenges). 
(not saying we should solve this for the current ramp, but IMO it is something we need to discuss more)

> The reason this is slightly messy here is because we don't yet know the detection workflow very well. Once we do, we can design a more granular standardized workflow (or even alternative workflows) with several files.

But the question is: will we ever know every possible workflow? And this thus also means that we impose one of the workflows we think are possible on the participant, possibly needlessly restricting him/her to try something else. 
Eg for this ramp, I was looking at the SSD deep net, and you were thinking about a workflow that first targets the mask. This are two completely different structures of deep models, and I don't think we should impose a kind of deep net architecture. If we don't, this means that there will be a lot of custom code inside one submission (most likely copy pasted from an example deep net online). 
(of course this all depends on what technical level we expect participants to be: only tweak small parts of proposed neural net, or submit whole new architectures)

**Comment by kegl** [comment]
We have to deal with uncertainty at this moment, which is hard. Standardization is the goal. It means constraints. We just don't yet know which ones.

I'm not sure though why putting workflows into rampwf is not scaleable. It's a very flat structure, and each new workflow comes with its own test. If its useful, it'll be reused. If not, it's just sitting there. If we can't get n_workflows << n_problems eventually, it means we failed at standardization.

**Comment by jorisvandenbossche** [comment]
IMO by always needing to add *everything* a ramp kit needs into ramp-workflow\*, you get the problems that: 1) you always need to merge before a ramp (otherwise the ramp cannot run), irregardless of the quality of the code, 2) therefore will not always think well about generalization before the ramp (no proper separation of concerns, mixing ramp-kit specific things with generalizable things), 3) end up with an unmaintainable mess.

Once it is in ramp-kits, it is also difficult to clean-up / fix things, as that makes previous kits unrunnable (but to solve that we probably just need better versioning). 
I think it will be often easier and result in better code to only after the ramp move those things that are general to ramp-workflow.

\* this is not strictly true, as you can of course put things inside problem.py (like a custom Prediction class)
